pkg/clusteragent/admission/mutate/autoinstrumentation/configuration_cache.go

```diff
@@ -0,0 +1,289 @@
+
+//go:build kubeapiserver
+
+// Package autoinstrumentation implements the webhook that injects APM libraries
+// into pods
+package autoinstrumentation
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type instrumentationConfiguration struct {
+	enabled            bool
+	enabledNamespaces  []string
+	disabledNamespaces []string
+}
+
+type enablementConfig struct {
+	configID          string
+	rcVersion         int
+	rcAction          string
+	env               *string
+	enabled           *bool
+	enabledNamespaces *[]string
+}
+
+type instrumentationConfigurationCache struct {
+	localConfiguration           *instrumentationConfiguration
+	currentConfiguration         *instrumentationConfiguration
+	configurationUpdatesQueue    chan Request
+	configurationUpdatesResponse chan Response
+	clusterName                  string
+	namespaceToConfigIDMap       map[string]string // maps the namespace with enabled instrumentation to Remote Enablement rule
+
+	mu               sync.RWMutex
+	orderedRevisions []int64
+	enabledRevisions map[int64]enablementConfig
+}
+
+func newInstrumentationConfigurationCache(
+	localEnabled *bool,
+	localEnabledNamespaces *[]string,
+	localDisabledNamespaces *[]string,
+	clusterName string,
+) *instrumentationConfigurationCache {
+	localConfig := newInstrumentationConfiguration(localEnabled, localEnabledNamespaces, localDisabledNamespaces)
+	currentConfig := newInstrumentationConfiguration(localEnabled, localEnabledNamespaces, localDisabledNamespaces)
+	reqChannel := make(chan Request, 10)
+	respChannel := make(chan Response, 10)
+
+	nsToRules := make(map[string]string)
+	if *localEnabled {
+		for _, ns := range *localEnabledNamespaces {
+			nsToRules[ns] = "local"
+		}
+	}
+
+	return &instrumentationConfigurationCache{
+		localConfiguration:           localConfig,
+		currentConfiguration:         currentConfig,
+		configurationUpdatesQueue:    reqChannel,
+		configurationUpdatesResponse: respChannel,
+		clusterName:                  clusterName,
+		namespaceToConfigIDMap:       nsToRules,
+
+		orderedRevisions: make([]int64, 0),
+		enabledRevisions: map[int64]enablementConfig{},
+	}
+}
+
+func (c *instrumentationConfigurationCache) start(stopCh <-chan struct{}) {
+	for {
+		select {
+		case req := <-c.configurationUpdatesQueue:
+			// if err := c.updateConfiguration(nil, nil); err != nil {
+			// 	log.Error(err.Error())
+			// }
+			resp := c.update(req)
+			c.configurationUpdatesResponse <- resp
+		case <-stopCh:
+			log.Info("Shutting down patcher")
+			return
+		}
+	}
+}
+
+func (c *instrumentationConfigurationCache) update(req Request) Response {
+
+	k8sClusterTargets := req.K8sTargetV2.ClusterTargets
+	var resp Response
+
+	switch req.Action {
+	case EnableConfig:
+		for _, target := range k8sClusterTargets {
+			clusterName := target.ClusterName
+
+			if c.clusterName == clusterName {
```

Comment:
Nit: I think it's better to write
```go
if c.clusterName != clusterName {
  continue
}
```

Less indentations needed and a bit cleaner.